tsdb/index: Refactor Reader tests

aknuds1 opened the pull request:
Refactor `tsdb/index.Reader` tests and a benchmark, as they have a lot in common. Pulled out of #13620, as the refactor became irrelevant there :)

I think the tests/benchmark become much more maintainable and easier to understand after this refactoring.

beorn7 reviewed (changes requested):
LGTM. Just a comment style suggestion, and of course, merge conflicts…

aknuds1 commented:
Thanks for the review @beorn7 - I fixed the merge conflicts (and refactored `TestReader_PostingsForLabelMatchingHonorsContextCancel` to use `createFileReader`), and applied your suggestion. PTAL :)

beorn7 commented:
Thank you, will merge now.
